File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph-dfs.java
class Solution {
    public static boolean dfs(int src, int dest, LinkedList<Integer>[] g, Set<Integer> visited){
        if(src == dest) return true;

        visited.add(src);

        boolean ans = false;

        for(int neighbour : g[src]){
            if(!visited.contains(neighbour)){
                ans = ans || dfs(neighbour, dest, g, visited);
            }
        }

        return ans;
    }

    public boolean validPath(int n, int[][] edges, int source, int destination) {
        LinkedList<Integer>[] g = new LinkedList[n];

        for(int i = 0; i < n; i++){
            g[i] = new LinkedList<>();
        }

        for(int i = 0; i < edges.length; i++){
            int u = edges[i][0];
            int v = edges[i][1];

            g[u].add(v);
            g[v].add(u);
        }

        Set<Integer> visited = new HashSet<>();

        return dfs(source, destination, g, visited);
    }
}

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph-dfs.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
function dfs(src, dest, graph, visited){
    if(src == dest) return true;

    let ans = false;
    visited.add(src);  // mark scr as visited

    for(const neighbour of graph[src]) {
        if(!visited.has(neighbour)) {
            // if we have not already visited the neighbour go
            ans = ans || dfs(neighbour, dest, graph, visited);
        }
    }
    return ans;
}

var validPath = function(n, edges, source, destination) {
    let graph = new Array(n);  // make graph like adjacency list 

    for(let i = 0; i < n; i++){
        graph[i] = [];  // for easier just used array but you should use LL for optimization
    }

    for(let i = 0; i < edges.length; i++){
        let u = edges[i][0];
        let v = edges[i][1];   // as graph is undirected if 0 -> 1 then 1 -> 0 also
        graph[u].push(v);
        graph[v].push(u);
    }

    return dfs(source, destination, graph, new Set())   // set for keeping track of visited edges
};